tee/tee.1
.Dd December 2008
.Dt TEE 1

.Sh NAME

.Nm tee
.Nd duplicate standard input

.Sh SYNOPSIS

.Nm tee
.Ar [-ai][file...]

.Sh DESCRIPTION

    The tee utility shall copy standard input to standard output, making a
copy in zero or more files. The tee utility shall not buffer output.

    If the -a option is not specified, output files shall be written (see
File Read, Write, and Creation.

.Sh OPTIONS

    The tee utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a
        Append the output to the files.
    -i
        Ignore the SIGINT signal.

.Sh OPERANDS

    The following operands shall be supported:

    file
        A pathname of an output file. Processing of at least 13 file operands
shall be supported.

.Sh STDIN

    The standard input can be of any type.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of tee:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default, except that if the -i option was specified, SIGINT shall be
ignored.

.Sh STDOUT

    The standard output shall be a copy of the standard input.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    If any file operands are specified, the standard input shall be copied to
each named file.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The standard input was successfully copied to all output files.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    If a write to any successfully opened file operand fails, writes to other
successfully opened file operands and standard output shall continue, but the
exit status shall be non-zero. Otherwise, the default actions specified in
Utility Description Defaults apply.

The following sections are informative.
.Sh APPLICATION USAGE

    The tee utility is usually used in a pipeline, to make a copy of the
output of some utility.

    The file operand is technically optional, but tee is no more useful than
cat when none is specified.

.Sh EXAMPLES

    Save an unsorted intermediate form of the data in a pipeline:

    ... | tee unsorted | sort > sorted

.Sh RATIONALE

    The buffering requirement means that tee is not allowed to use ISO C
standard fully buffered or line-buffered writes. It does not mean that tee
has to do 1-byte reads followed by 1-byte writes.

    It should be noted that early versions of BSD ignore any invalid options
and accept a single '-' as an alternative to -i. They also print a message if
unable to open a file:

    "tee: cannot access %s\n", <pathname>

    Historical implementations ignore write errors. This is explicitly not
permitted by this volume of IEEE Std 1003.1-2001.

    Some historical implementations use O_APPEND when providing append mode;
others use the lseek() function to seek to the end-of-file after opening the
file without O_APPEND. This volume of IEEE Std 1003.1-2001 requires
functionality equivalent to using O_APPEND; see File Read, Write, and
Creation.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    Introduction, cat, the System Interfaces volume of IEEE Std 1003.1-2001,
lseek()



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



tee.c
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/select.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <libpu.h>

#define PFX "tee: "

static const char doc[] =
N_("tee - read from standard input and write to standard output and files");

static struct argp_option options[] = {
	{ "append", 'a', NULL, 0,
	  N_("Append the output to the files") },
	{ "ignore-interrupts", 'i', NULL, 0,
	  N_("Ignore the SIGINT signal") },
	{ }
};

static int tee_post_walk(struct walker *w);
static int tee_init(struct walker *w, int argc, char **argv);
static int build_flist_actor(struct walker *w, const char *fn, const struct stat *lst);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, file_args_doc, doc };

static struct walker walker;

#define ERRSTR_NOMEM PFX "out of memory\n"
#define REALLY_FREE
#define TEE_BUF_SZ 8192

static char buf[TEE_BUF_SZ];
static bool opt_append;
static bool opt_sigint;

struct tee_flist {
	const char *fn;
	int fd;
	struct tee_flist *next;
	unsigned int skip;
};

static struct tee_flist *flist_root;

static int free_flist(void)
{
	struct tee_flist *tmp;
	int err = 0;

	while (flist_root) {
		tmp = flist_root;
		flist_root = flist_root->next;

		if (close(tmp->fd) < 0) {
			perror(tmp->fn);
			err |= 1;
		}

#ifdef REALLY_FREE
		free(tmp);
#endif
	}

	flist_root = NULL;

	return err;
}

static int tee_output_bytes(const char *buf, ssize_t buflen)
{
	struct tee_flist *tmp = flist_root;
	ssize_t wrc, to_write;
	const char *s;
	int err = 0;

	while (tmp) {
		if (!tmp->skip) {
			s = buf;
			to_write = buflen;

			while (to_write > 0) {
				wrc = write(tmp->fd, s, to_write);
				if (wrc < 1) {
					perror(tmp->fn);
					tmp->skip = 1;
					err = 1;
					break;
				}

				s += wrc;
				to_write -= wrc;
			}
		}

		tmp = tmp->next;
	}

	return err;
}

static int tee_output(void)
{
	int rc, err = 0;
	fd_set rd_set, err_set;
	ssize_t bread;

	rc = fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
	if (rc < 0) {
		perror(STDIN_NAME);
		err = 1;
		goto out;
	}

	FD_ZERO(&rd_set);
	FD_ZERO(&err_set);
	while (1) {
		FD_SET(STDIN_FILENO, &rd_set);
		FD_SET(STDIN_FILENO, &err_set);

		rc = select(STDIN_FILENO + 1, &rd_set, NULL, &err_set, NULL);
		if (rc < 1) {
			perror(STDIN_NAME);
			break;
		}

		bread = read(STDIN_FILENO, buf, sizeof(buf));
		if (bread == 0)
			break;
		if (bread < 0) {
			perror(STDIN_NAME);
			rc = 1;
			break;
		}

		err |= tee_output_bytes(buf, bread);
	}

out:
	err |= free_flist();

	return err;
}

static int build_flist_actor(struct walker *w, const char *fn, const struct stat *lst)
{
	struct tee_flist *new_ent, *tmp;
	int open_flags;
	int fd;

	if (opt_append)
		open_flags = O_WRONLY | O_CREAT | O_APPEND;
	else
		open_flags = O_WRONLY | O_CREAT | O_TRUNC;

	fd = open(fn, open_flags, 0666);
	if (fd < 0) {
		perror(fn);
		return 1;
	}

	new_ent = (struct tee_flist *) xmalloc(sizeof(*new_ent));
	new_ent->fn = fn;
	new_ent->fd = fd;
	new_ent->next = NULL;

	tmp = flist_root;
	while (tmp->next)
		tmp = tmp->next;
	tmp->next = new_ent;

	return 0;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN
	PU_OPT_SET('a', append)
	PU_OPT_SET('i', sigint)
	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

static int tee_init(struct walker *w, int argc, char **argv)
{
	struct tee_flist *new_ent;

	pu_init();

	new_ent = (struct tee_flist *) xmalloc(sizeof(*new_ent));
	new_ent->fn = STDOUT_NAME;
	new_ent->fd = STDOUT_FILENO;
	new_ent->next = NULL;
	flist_root = new_ent;

	return 0;
}

static int tee_post_walk(struct walker *w)
{
	if (opt_sigint && signal(SIGINT, SIG_IGN) == SIG_ERR) {
		fprintf(stderr, PFX "cannot ignore SIGINT\n");
		return 1;
	}

	return tee_output();
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.init			= tee_init;
	walker.post_walk		= tee_post_walk;
	walker.cmdline_arg		= build_flist_actor;
	return walk(&walker, argc, argv);
}
